src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.OneNote/Main.cs

```diff
@@ -83,28 +92,52 @@ public void Init(PluginInitContext context)
         /// <returns>A filtered list, can be empty when nothing was found</returns>
         public List<Result> Query(Query query)
         {
-            if (!_oneNoteInstalled || query is null || string.IsNullOrWhiteSpace(query.Search))
+            if (!_oneNoteInstalled || query is null || string.IsNullOrWhiteSpace(query.Search) || cache is null)
             {
                 return new List<Result>(0);
             }
 
-            var pages = OneNoteProvider.FindPages(query.Search);
+            // If there's cached results for this query, return immediately, otherwise wait for delayedExecution.
+            var results = cache.Get<List<Result>>(query.Search);
+            return results ?? Query(query, false);
+        }
+
+        /// <summary>
+        /// Return a filtered list, based on the given query
+        /// </summary>
+        /// <param name="query">The query to filter the list</param>
+        /// <param name="delayedExecution">False if this is the first pass through plugins, true otherwise. Slow plugins should run delayed.</param>
+        /// <returns>A filtered list, can be empty when nothing was found</returns>
+        public List<Result> Query(Query query, bool delayedExecution)
+        {
+            if (!delayedExecution || !_oneNoteInstalled || query is null || string.IsNullOrWhiteSpace(query.Search) || cache is null)
+            {
+                return new List<Result>(0);
+            }
 
-            return pages.Select(p => new Result
+            // Get results from cache if they already exist for this query, otherwise query OneNote. Results will be cached for 1 day.
+            var results = cache.GetOrAdd(query.Search, () =>
             {
-                IcoPath = _iconPath,
-                Title = p.Name,
-                QueryTextDisplay = p.Name,
-                SubTitle = @$"{p.Notebook.Name}\{p.Section.Name}",
-                Action = (_) =>
+                var pages = OneNoteProvider.FindPages(query.Search);
```

Comment:
Can we cache the page data instead of the results? If delay in PT is disabled there is a query for each character.¹ Or is it different on delayed execution?

¹) So we get a cache for a, b, c, ... which makes no sense and only fills memory.

Reply:
Please see this comment: https://github.com/microsoft/PowerToys/issues/18846#issuecomment-115954808

And maybe you should reduce the cache time to avoid outdated cache data.

Reply:
Responded to comment in the issue, thanks!